jni/src/Foreign/JNI/Types.hs

```diff
@@ -220,7 +227,7 @@ instance Storable JValue where
   poke p (JLong x) = poke (castPtr p) x
   poke p (JFloat x) = poke (castPtr p) x
   poke p (JDouble x) = poke (castPtr p) x
-  poke p (JObject x) = poke (castPtr p :: Ptr (J a)) x
+  poke p (JObject (J x)) = withForeignPtr x (poke (castPtr p))
```

Comment:
I didn't change this part of the patch, because I didn't see a problem with it. But there might indeed be a race here.

We can't tell the user not to `poke`. It's what `call*Method` and therefore what `call` does under the hood for all arguments. Problem is we now have a circular dependency between `Foreign.JNI` and `Foreign.JNI.Types`...